src/gwt/src/org/rstudio/core/client/widget/InfoBar.java

```diff
@@ -93,6 +95,44 @@ public int getHeight()
       return 19;
    }
    
+   public void showRequiredPackagesMissingWarning(List<String> packages,
+                                                  CommandWithArg<Boolean> onPackageInstallCompleted,
+                                                  Command onDismiss)
+   {
+      String message;
+      
+      int n = packages.size();
+      if (n == 1)
+      {
+         message = "Package " + packages.get(0) + " required but is not installed.";
+      }
+      else if (n == 2)
+      {
+         message = "Packages " + packages.get(0) + " and " + packages.get(1) + " required but are not installed.";
+      }
+      else if (n == 3)
+      {
+         message = "Packages " + packages.get(0) + ", " + packages.get(1) + ", and " + packages.get(2) + " required but are not installed.";
+      }
+      else
+      {
+         message = "Packages " + packages.get(0) + ", " + packages.get(1) + ", and " + (n - 2) + " others are required but not installed.";
+      }
+      
+      label_.setText(message);
+      
+      if (labelRight_.getWidgetCount() == 0)
+      {
+         labelRight_.add(label("Install", () -> {
+            RStudioGinjector.INSTANCE.getDependencyManager().installPackages(packages, onPackageInstallCompleted);
+         }));
+         
```

Comment:
I think we need something between "install right now" and "don't ever show this again". Can we add *Hide*? (I can't remember whether the warning bar has dismissal mechanism built in; disregard if it does!)

Reply:
The warning bar does indeed have a little red [X] at the far right for dismissal :)